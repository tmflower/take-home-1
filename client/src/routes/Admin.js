// import { useState } from "react";
import { Box, Typography, Table, TableHead, TableBody, TableRow, TableCell } from "@mui/material";
import theme from "../theme";

export function Admin({users}) {

    return (
        <Box sx={{margin: '8rem'}}>
            <Typography variant="h2" sx={{margin: '6rem 0 0 0', display: 'flex', flexDirection: 'column', alignItems: 'center'}}>Admin</Typography>
            <Table>
                <TableHead>
                    <TableRow>
                        <TableCell sx={{color: theme.typography.primary.color, fontSize: '1.5rem'}}>ID</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color, fontSize: '1.5rem'}}>First name</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color, fontSize: '1.5rem'}}>Last name</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color, fontSize: '1.5rem'}}>Email</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color, fontSize: '1.5rem'}}>Status</TableCell>
                    </TableRow>
                </TableHead>
                <TableBody>                    
                    {users.map((user, i) =>
                    <TableRow key={i}>                        
                        <TableCell sx={{color: theme.typography.primary.color}}>{user.id}</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color}}>{user.firstName}</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color}}>{user.lastName}</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color}}>{user.email}</TableCell>
                        <TableCell sx={{color: theme.typography.primary.color}}>{user.state}</TableCell>
                    </TableRow>)}
                </TableBody>
            </Table>
        </Box>
    )
}