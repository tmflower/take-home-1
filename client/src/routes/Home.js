import { Box, Button, Typography } from "@mui/material";
import { NavLink } from "react-router-dom";

export function Home() {
    return (
        <Box sx={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            margin: '6rem'}}>
            <Typography variant="h1">Welcome to Yodlr!</Typography>
            <Typography variant="h3">Who are you?</Typography>
            <Box sx={{display: 'flex', flexDirection: 'row', }}>                    
                <NavLink to="/signup" className="link">
                    <Button variant="contained" sx={{margin: '2rem', backgroundColor: "chocolate"}}>User</Button>
                </NavLink>                
                <NavLink to="/admin">
                    <Button variant="contained" sx={{margin: '2rem', backgroundColor: "chocolate"}}>Admin</Button>
                </NavLink>
                
            </Box>
        </Box>
    )
}