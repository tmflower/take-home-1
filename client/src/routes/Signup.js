import { useState } from "react";
import axios from "axios";

export function Signup({ setUserAdded }) {
    const [message, setMessage] = useState('')
    const [formData, setFormData] = useState({
        firstName: '',
        lastName: '',
        email: ''
    });

    const { firstName, lastName, email } = formData;

    const handleChange = (evt) => {
        const { name, value } = evt.target;
        setFormData(formData => ({...formData, [name]: value}));
    };

    const handleSubmit = async(evt) => {
        evt.preventDefault();
        try { 
            await axios.post('http://localhost:3001/users', { firstName, lastName, email });
            setMessage(`Welcome to Yodlr, ${firstName}!`);
            setUserAdded(true);
        }
        catch(err) {
            console.error(err);
        }
    }
    return (
        <div>
            <h1>Yodlr Registration Portal</h1>
            {message ? <p>{message}</p> : null}
            <form onSubmit={handleSubmit}>
                <label htmlFor="firstName">
                    First name:
                    <input
                        type="text"
                        name="firstName"
                        value={firstName}
                        id="firstName"
                        onChange={handleChange}>
                    </input>                    
                </label>
                <label htmlFor="lastName">
                    Last name:
                    <input
                        type="text"
                        name="lastName"
                        value={lastName}
                        id="lastName"
                        onChange={handleChange}>
                    </input>
                </label>
                <label htmlFor="email">
                    Email:
                    <input
                        type="email"
                        name="email"
                        value={email}
                        id="email"
                        onChange={handleChange}>
                    </input>
                </label>
            <button type="submit">Submit</button>
            </form>
        </div>
    )
}