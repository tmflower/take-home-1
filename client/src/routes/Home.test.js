import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { Home } from './Home';
import { Signup } from './Signup';
import { Admin } from './Admin';
import userEvent from "@testing-library/user-event";

const users = [
    { "id": 1, "email": "kyle@getyodlr.com", "firstName": "Kyle", "lastName": "White", "state": "active"},
    { "id": 2, "email": "jane@getyodlr.com", "firstName": "Jane", "lastName": "Stone", "state": "active"},
    { "id": 3, "email": "lilly@getyodlr.com", "firstName": "Lilly", "lastName": "Smith", "state": "pending"},
    { "id": 4, "email": "fred@getyodlr.com", "firstName": "Fred", "lastName": "Miles", "state": "pending"},
    { "id": 5, "email": "alex@getyodlr.com", "firstName": "Alexandra", "lastName": "Betts", "state": "pending"}
    ]

test('renders Home without crashing', () => {
    render(
    <MemoryRouter>
        <Home />
    </MemoryRouter>);
});

test('matches snapshot', () => {
    const { asFragment } =
    render(
        <MemoryRouter>
            <Home />
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test('renders Home correctly', () => {
    render(
        <MemoryRouter>
            <Home />
        </MemoryRouter>
    );
    const h1 = screen.getByText('Welcome to Yodlr!');
    expect(h1).toBeInTheDocument();
});

test('handles button click for user', async () => {
    render(
        <MemoryRouter>
            <Routes>
                <Route path="/signup" element={<Signup />} />
                <Route path="/" element={<Home />} />
            </Routes>
        </MemoryRouter>
    );

    const headingBefore = screen.queryByText('Yodlr Registration Portal');
    expect(headingBefore).not.toBeInTheDocument();

    userEvent.click(screen.queryByRole('link', { name: 'User' }));
    
    await waitFor(() => {
        const headingAfter = screen.queryByText('Yodlr Registration Portal');
        expect(headingAfter).toBeInTheDocument();
    });
});

test('handles button click for admin', async () => {
    render(
        <MemoryRouter>
            <Routes>
                <Route path="/admin" element={<Admin users={users} />} />
                <Route path="/" element={<Home />} />
            </Routes>
        </MemoryRouter>
    );

    const headingBefore = screen.queryByText('Admin Portal');
    expect(headingBefore).not.toBeInTheDocument();

    userEvent.click(screen.queryByRole('link', { name: 'Admin' }))
    
    await waitFor(() => {
        const headingAfter = screen.queryByText('Admin Portal');
        expect(headingAfter).toBeInTheDocument();
    });
});
