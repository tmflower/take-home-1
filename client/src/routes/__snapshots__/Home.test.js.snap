// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1d27sqt"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
    >
      Welcome to Yodlr!
    </h1>
    <h3
      class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
    >
      Who are you?
    </h3>
    <div
      class="MuiBox-root css-1xhj18k"
    >
      <a
        class="link"
        href="/signup"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-usvop9-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          User
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
      <a
        class=""
        href="/admin"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-usvop9-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Admin
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
